[PyTorch Mobile][Forward/backward compatibility] Number of arguments for operators

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56845 [PyTorch Mobile][Forward/backward compatibility] Number of arguments for operators**

Handle forward/backward compatibility caused by added default arguments in mobile. As an example,

In older version, operator aten::foo's schema is
```
foo(Tensor a, Tensor b) -> Tensor
```
In the new version, the schema is updated to
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```

## Model file
Serialize the number of specified arguments to each operator into the bytecode operator table. Before the operator table contains operator name and overload name:
```
('operators', (('aten::foo', ''),))
```
Now the number of specified arguments is added:
```
# bytecode version 6
('operators', (('aten::foo', '', 2),))
```
where "2" means the number of specified arguments. 

Since there's bytecode schema change, the bytecode version number is bumped. This PR is to be landed after #56002 , where the version number is bumped from 4 to 5. This PR bumps the version number from 5 to 6. 

## Runtime and backward compatibility
When the operator is found (either jit or c10), we have the OperatorHandle, where the operator schema can be accessed by 
```
op.value().schema().arguments()
```
Adaptation is implemented to handle backward compatibility. For the example above, the new runtime holds the updated schema: 
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```
Whereas the model file carries
```
(('aten::foo', ''), 2)
```
We can implement a wrapper around the original function pointer to push the default argument to the stack.

## Deliver time and forward compatibility
At model delivery time, two checks can be done:
### Operator check
Two APIs to be provided:
* Runtime: An API to get a runtime’s ops and their schemas (i.e. the # of args). D27920185(WIP)
* Model: An API to get a model’s ops and their schema requirements (i.e. the # of args required).

The APIs can be used to check 
* runtime.ops() is a superset of model.ops()
* for each op in model.ops() validate their schemas are compatible with those in runtime.ops() -- i.e. the # args required in a model op are <= # args in the runtime op.

Note that only root ops in the model needs to be checked here. For transient ops it's not necessary. For example, if a root op, "aten::root" calls "aten::foo", it's "aten::root"'s responsibility to adapt to "aten::foo"'s change, or "aten::root" itself needs to be updated too. 
### Bytecode version backport (PR coming)
When delivering a model with bytecode v6, if the runtime only works with bytecode v5 and lower, backport is needed. 
* The number of arguments is removed from the operator table
* The bytecode version is changed from 6 to 5

Note that this backport is a pure format change, it does not guarantee the backported model always runs in old runtime. The operator check mentioned before should be done first, before it’s back ported to v5.

Differential Revision: [D27986544](https://our.internmc.facebook.com/intern/diff/D27986544)

iseeyuan committed:
[PyTorch Mobile][Forward/backward compatibility] Number of arguments for operators

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch Mobile][Forward/backward compatibility] Number of arguments for operators"


Handle forward/backward compatibility caused by added default arguments in mobile. As an example,

In older version, operator aten::foo's schema is
```
foo(Tensor a, Tensor b) -> Tensor
```
In the new version, the schema is updated to
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```

## Model file
Serialize the number of specified arguments to each operator into the bytecode operator table. Before the operator table contains operator name and overload name:
```
('operators', (('aten::foo', ''),))
```
Now the number of specified arguments is added:
```
# bytecode version 6
('operators', (('aten::foo', '', 2),))
```
where "2" means the number of specified arguments. 

Since there's bytecode schema change, the bytecode version number is bumped. This PR is to be landed after #56002 , where the version number is bumped from 4 to 5. This PR bumps the version number from 5 to 6. 

## Runtime and backward compatibility
When the operator is found (either jit or c10), we have the OperatorHandle, where the operator schema can be accessed by 
```
op.value().schema().arguments()
```
Adaptation is implemented to handle backward compatibility. For the example above, the new runtime holds the updated schema: 
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```
Whereas the model file carries
```
(('aten::foo', ''), 2)
```
We can implement a wrapper around the original function pointer to push the default argument to the stack.

## Deliver time and forward compatibility
At model delivery time, two checks can be done:
### Operator check
Two APIs to be provided:
* Runtime: An API to get a runtime’s ops and their schemas (i.e. the # of args). D27920185(WIP)
* Model: An API to get a model’s ops and their schema requirements (i.e. the # of args required).

The APIs can be used to check 
* runtime.ops() is a superset of model.ops()
* for each op in model.ops() validate their schemas are compatible with those in runtime.ops() -- i.e. the # args required in a model op are <= # args in the runtime op.

Note that only root ops in the model needs to be checked here. For transient ops it's not necessary. For example, if a root op, "aten::root" calls "aten::foo", it's "aten::root"'s responsibility to adapt to "aten::foo"'s change, or "aten::root" itself needs to be updated too. 
### Bytecode version backport (PR coming)
When delivering a model with bytecode v6, if the runtime only works with bytecode v5 and lower, backport is needed. 
* The number of arguments is removed from the operator table
* The bytecode version is changed from 6 to 5

Note that this backport is a pure format change, it does not guarantee the backported model always runs in old runtime. The operator check mentioned before should be done first, before it’s back ported to v5.



[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -642,8 +651,8 @@ void export_opnames(const script::Module& m, std::set<std::string>& opnames) {
       TORCH_CHECK(
-          op_item.size() == 2,
-          "There should be two parts in an operator name.");
+          op_item.size() >= 2,
```
is this because the third one can be num_speficied_args?

iseeyuan replied:
Right. Before it has just name and overload. New the num_speficied_args is added.

raziel replied:
Same comment about documenting the "parts".

iseeyuan committed:
Update on "[PyTorch Mobile][Forward/backward compatibility] Number of arguments for operators"


Handle forward/backward compatibility caused by added default arguments in mobile. As an example,

In older version, operator aten::foo's schema is
```
foo(Tensor a, Tensor b) -> Tensor
```
In the new version, the schema is updated to
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```

## Model file
Serialize the number of specified arguments to each operator into the bytecode operator table. Before the operator table contains operator name and overload name:
```
('operators', (('aten::foo', ''),))
```
Now the number of specified arguments is added:
```
# bytecode version 6
('operators', (('aten::foo', '', 2),))
```
where "2" means the number of specified arguments. 

Since there's bytecode schema change, the bytecode version number is bumped. This PR is to be landed after #56002 , where the version number is bumped from 4 to 5. This PR bumps the version number from 5 to 6. 

## Runtime and backward compatibility
When the operator is found (either jit or c10), we have the OperatorHandle, where the operator schema can be accessed by 
```
op.value().schema().arguments()
```
Adaptation is implemented to handle backward compatibility. For the example above, the new runtime holds the updated schema: 
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```
Whereas the model file carries
```
(('aten::foo', ''), 2)
```
We can implement a wrapper around the original function pointer to push the default argument to the stack.

## Deliver time and forward compatibility
At model delivery time, two checks can be done:
### Operator check
Two APIs to be provided:
* Runtime: An API to get a runtime’s ops and their schemas (i.e. the # of args). D27920185(WIP)
* Model: An API to get a model’s ops and their schema requirements (i.e. the # of args required).

The APIs can be used to check 
* runtime.ops() is a superset of model.ops()
* for each op in model.ops() validate their schemas are compatible with those in runtime.ops() -- i.e. the # args required in a model op are <= # args in the runtime op.

Note that only root ops in the model needs to be checked here. For transient ops it's not necessary. For example, if a root op, "aten::root" calls "aten::foo", it's "aten::root"'s responsibility to adapt to "aten::foo"'s change, or "aten::root" itself needs to be updated too. 
### Bytecode version backport (PR coming)
When delivering a model with bytecode v6, if the runtime only works with bytecode v5 and lower, backport is needed. 
* The number of arguments is removed from the operator table
* The bytecode version is changed from 6 to 5

Note that this backport is a pure format change, it does not guarantee the backported model always runs in old runtime. The operator check mentioned before should be done first, before it’s back ported to v5.

Differential Revision: [D27986544](https://our.internmc.facebook.com/intern/diff/D27986544)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1160,26 +1160,25 @@ struct MobileCodeImpl : CodeImpl {
+          ? op.schema().name() + "." + op.schema().overload_name()
+          : op.schema().name();
+      auto num_include = node->inputs().size();
+      // make sure we only do this for mobile code
+      if (op_to_num_specified_args_.find(unique_op_name) !=
+          op_to_num_specified_args_.end()) {
+        num_include = op_to_num_specified_args_[unique_op_name];
+      }
```
@tugsbayasgalan 
1. I feel this comment doesn't exactly match the implementation. The condition checks whether an op has trimmed arguments, not whether an op is supported by mobile. Since this emitOperator is implemented only for `MobileCodeImpl`, we can assume all ops are for mobile I think.
2. We can use same iterator that's returned from `find` to look up `num_include` to save one map look-up

raziel replied:
yeah, let's make it more specific about whether an op has trimmed arguments.

tugsbayasgalan replied:
Yeah will remove it. good catch

gmagogsfm commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -198,9 +198,18 @@ std::pair<IValue, c10::optional<IValue>> getFunctionTuple(
   operators.reserve(opnames.size());
   for (const auto& opname : opnames) {
-    operators.emplace_back(Tup({opname.name, opname.overload_name}));
+    auto unique_name = opname.overload_name != ""
+        ? opname.name + "." + opname.overload_name
+        : opname.name;
```
Can this be refactored into a utility function so that it is shared by all places that need a unique name to prevent future divergence?

gmagogsfm commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -943,6 +943,174 @@ TEST(LiteInterpreterTest, OpNameExportFetchRootOperators) {
+  // The second argument is specified, but the value is the same as the default
+  // value. It's treated as "not specified" since the value can be fetched from
+  // schema.
+  Module m("m");
+  m.define(R"(
+    def forward(self, input):
+      return torch.linalg_pinv(input, 1e-15)
+  )");
+  torch::jit::MobileCode code(m.get_method("forward").graph(), "forward");
+  auto arg_nums = code.op_to_num_specified_args();
+  ASSERT_EQ(arg_nums.size(), 1);
+  ASSERT_EQ(arg_nums["aten::linalg_pinv"], 1);
+  std::vector<torch::jit::IValue> inputs;
+  inputs.push_back(torch::rand({28, 28}));
+  testLiteModuleCompareResultTensors(m, inputs);
```
Super nit pick:
This feels more like an independent scenario (and an important one), so maybe it should be its own test case.

gmagogsfm commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -943,6 +943,174 @@ TEST(LiteInterpreterTest, OpNameExportFetchRootOperators) {
+  }
+
+  std::vector<torch::jit::IValue> inputs;
+  inputs.push_back(torch::rand({28, 28}));
```
Nit;

IMHO, something like torch arange would be faster and more deterministic than `rand`

gmagogsfm reviewed (approved):
Looks really good, just have some minor comments.

raziel commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -171,11 +171,16 @@ std::unordered_set<std::string> BytecodeDeserializer::
     auto op_item = op.toTuple()->elements();
     TORCH_CHECK(
-        op_item.size() == 2, "There should be two parts in an operator name.");
+        op_item.size() >= 2,
```
Maybe we should document here what parts are expected (I wish we have a better schema definition so we didn't have to but since we don't then we should describe it at least).


Also, in the future if we ever make the min version is >= that 6 we could remove this check, right? Just curious.

iseeyuan replied:
Right. If we have the byte code version check >= 6, this check can be removed.

tugsbayasgalan commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -642,8 +651,8 @@ void export_opnames(const script::Module& m, std::set<std::string>& opnames) {
-          op_item.size() == 2,
-          "There should be two parts in an operator name.");
+          op_item.size() >= 2,
+          "There should at least two parts in an operator name.");
```
I agree. There can't be more than 3 parts right?

iseeyuan committed:
Update on "[PyTorch Mobile][Forward/backward compatibility] Number of arguments for operators"


Handle forward/backward compatibility caused by added default arguments in mobile. As an example,

In older version, operator aten::foo's schema is
```
foo(Tensor a, Tensor b) -> Tensor
```
In the new version, the schema is updated to
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```

## Model file
Serialize the number of specified arguments to each operator into the bytecode operator table. Before the operator table contains operator name and overload name:
```
('operators', (('aten::foo', ''),))
```
Now the number of specified arguments is added:
```
# bytecode version 6
('operators', (('aten::foo', '', 2),))
```
where "2" means the number of specified arguments. 

Since there's bytecode schema change, the bytecode version number is bumped. This PR is to be landed after #56002 , where the version number is bumped from 4 to 5. This PR bumps the version number from 5 to 6. 

## Runtime and backward compatibility
When the operator is found (either jit or c10), we have the OperatorHandle, where the operator schema can be accessed by 
```
op.value().schema().arguments()
```
Adaptation is implemented to handle backward compatibility. For the example above, the new runtime holds the updated schema: 
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```
Whereas the model file carries
```
(('aten::foo', ''), 2)
```
We can implement a wrapper around the original function pointer to push the default argument to the stack.

## Deliver time and forward compatibility
At model delivery time, two checks can be done:
### Operator check
Two APIs to be provided:
* Runtime: An API to get a runtime’s ops and their schemas (i.e. the # of args). D27920185(WIP)
* Model: An API to get a model’s ops and their schema requirements (i.e. the # of args required).

The APIs can be used to check 
* runtime.ops() is a superset of model.ops()
* for each op in model.ops() validate their schemas are compatible with those in runtime.ops() -- i.e. the # args required in a model op are <= # args in the runtime op.

Note that only root ops in the model needs to be checked here. For transient ops it's not necessary. For example, if a root op, "aten::root" calls "aten::foo", it's "aten::root"'s responsibility to adapt to "aten::foo"'s change, or "aten::root" itself needs to be updated too. 
### Bytecode version backport (PR coming)
When delivering a model with bytecode v6, if the runtime only works with bytecode v5 and lower, backport is needed. 
* The number of arguments is removed from the operator table
* The bytecode version is changed from 6 to 5

Note that this backport is a pure format change, it does not guarantee the backported model always runs in old runtime. The operator check mentioned before should be done first, before it’s back ported to v5.

Differential Revision: [D27986544](https://our.internmc.facebook.com/intern/diff/D27986544)

[ghstack-poisoned]

dhruvbird commented:
> Since there's bytecode schema change, the bytecode version number is bumped. This PR is to be landed after #56002 , where the version number is bumped from 4 to 5. This PR bumps the version number from 5 to 6.

@iseeyuan I'm wondering if it makes sense to land this before the de-dup constant PR since this one is likely trivially backward compatible - i.e. it's easier to handle one fewer value in the tuple in the code code. Old code probably just ignores it.

dhruvbird commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -33,7 +33,7 @@ void Function::append_instruction(OpCode op, int X, int N) {
     const std::string& name,
     const std::string& overload_name,
-    int64_t model_version) {
+    int64_t num_specified_args) {
```
This change seems like it could go unnoticed. Does it make sense to add one more argument and see what breaks? If we add and remove an `int64_t` argument, then existing call sites will still build successfully (in case we missed fixing some of them up).

raziel replied:
It is a replacement for a field with the same semantics but a clearer name.

It's a valid concern, but definitely we should't changing the actual committed code's signature just to verify we didn't miss refactoring somewhere. I would expect @iseeyuan already checked all necessary places (probably doing what you suggest).

iseeyuan replied:
The only call site of this function is in mobile/import.cpp. Also I think it's a good decoupling from model version. The model version of 3 check is too special.

dhruvbird replied:
@raziel Yep! I meant locally (not for checking in this change) - thanks for clarifying that!

dhruvbird commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -171,11 +171,18 @@ std::unordered_set<std::string> BytecodeDeserializer::
   for (const auto& op : ops_list) {
     auto op_item = op.toTuple()->elements();
     TORCH_CHECK(
-        op_item.size() == 2, "There should be two parts in an operator name.");
```
Ignore my previous comment about existing runtime being able to load newer models. I didn't realize that we explicitly check for number of elements `== 2`.

iseeyuan replied:
I updated the code from `>= 2` to `==2 || == 3`. I think I need to change it back to `>=2` for future extension. If the extension (for example, the 4th field) is optional, `>=2` would not throw error, but neglect the extended fields. It can be useful for forward compatibility.

dhruvbird commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -198,9 +198,16 @@ std::pair<IValue, c10::optional<IValue>> getFunctionTuple(
   for (const auto& opname : opnames) {
-    operators.emplace_back(Tup({opname.name, opname.overload_name}));
+    auto unique_name = c10::toString(opname);
+    int num_args = -1;
```
Under what cases do we not expect to find an operator schema? Should this be an error if we always expect to find an operator schema?

iseeyuan replied:
`op_to_specified_args` does not include operator with varargs. For those operators we don't do any default argument pushing. I'll add a comment on this.

kimishpatel replied:
Which are the example operators with varargs? Presumably these are different from the ones with default args?

iseeyuan replied:
Among > 2000 ops, there are 10s vararg operators where the number of argument varies. Those ops are different from default args and goes through OPN instruction.

dhruvbird commented:
> > Since there's bytecode schema change, the bytecode version number is bumped. This PR is to be landed after #56002 , where the version number is bumped from 4 to 5. This PR bumps the version number from 5 to 6.
> 
> @iseeyuan I'm wondering if it makes sense to land this before the de-dup constant PR since this one is likely trivially backward compatible - i.e. it's easier to handle one fewer value in the tuple in the code code. Old code probably just ignores it.

Ignore the comment above - didn't realize that the existing code checks if the number of elements in the tuple `== 2`.

iseeyuan committed:
Update on "[PyTorch Mobile][Forward/backward compatibility] Number of arguments for operators"


Handle forward/backward compatibility caused by added default arguments in mobile. As an example,

In older version, operator aten::foo's schema is
```
foo(Tensor a, Tensor b) -> Tensor
```
In the new version, the schema is updated to
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```

## Model file
Serialize the number of specified arguments to each operator into the bytecode operator table. Before the operator table contains operator name and overload name:
```
('operators', (('aten::foo', ''),))
```
Now the number of specified arguments is added:
```
# bytecode version 6
('operators', (('aten::foo', '', 2),))
```
where "2" means the number of specified arguments. 

Since there's bytecode schema change, the bytecode version number is bumped. This PR is to be landed after #56002 , where the version number is bumped from 4 to 5. This PR bumps the version number from 5 to 6. 

## Runtime and backward compatibility
When the operator is found (either jit or c10), we have the OperatorHandle, where the operator schema can be accessed by 
```
op.value().schema().arguments()
```
Adaptation is implemented to handle backward compatibility. For the example above, the new runtime holds the updated schema: 
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```
Whereas the model file carries
```
(('aten::foo', ''), 2)
```
We can implement a wrapper around the original function pointer to push the default argument to the stack.

## Deliver time and forward compatibility
At model delivery time, two checks can be done:
### Operator check
Two APIs to be provided:
* Runtime: An API to get a runtime’s ops and their schemas (i.e. the # of args). D27920185(WIP)
* Model: An API to get a model’s ops and their schema requirements (i.e. the # of args required).

The APIs can be used to check 
* runtime.ops() is a superset of model.ops()
* for each op in model.ops() validate their schemas are compatible with those in runtime.ops() -- i.e. the # args required in a model op are <= # args in the runtime op.

Note that only root ops in the model needs to be checked here. For transient ops it's not necessary. For example, if a root op, "aten::root" calls "aten::foo", it's "aten::root"'s responsibility to adapt to "aten::foo"'s change, or "aten::root" itself needs to be updated too. 
### Bytecode version backport (PR coming)
When delivering a model with bytecode v6, if the runtime only works with bytecode v5 and lower, backport is needed. 
* The number of arguments is removed from the operator table
* The bytecode version is changed from 6 to 5

Note that this backport is a pure format change, it does not guarantee the backported model always runs in old runtime. The operator check mentioned before should be done first, before it’s back ported to v5.

Differential Revision: [D27986544](https://our.internmc.facebook.com/intern/diff/D27986544)

[ghstack-poisoned]

raziel reviewed (approved):
Thanks!

Commit:
Update on "[PyTorch Mobile][Forward/backward compatibility] Number of arguments for operators"


Handle forward/backward compatibility caused by added default arguments in mobile. As an example,

In older version, operator aten::foo's schema is
```
foo(Tensor a, Tensor b) -> Tensor
```
In the new version, the schema is updated to
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```

## Model file
Serialize the number of specified arguments to each operator into the bytecode operator table. Before the operator table contains operator name and overload name:
```
('operators', (('aten::foo', ''),))
```
Now the number of specified arguments is added:
```
# bytecode version 6
('operators', (('aten::foo', '', 2),))
```
where "2" means the number of specified arguments. 

Since there's bytecode schema change, the bytecode version number is bumped. This PR is to be landed after #56002 , where the version number is bumped from 4 to 5. This PR bumps the version number from 5 to 6. 

## Runtime and backward compatibility
When the operator is found (either jit or c10), we have the OperatorHandle, where the operator schema can be accessed by 
```
op.value().schema().arguments()
```
Adaptation is implemented to handle backward compatibility. For the example above, the new runtime holds the updated schema: 
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```
Whereas the model file carries
```
(('aten::foo', ''), 2)
```
We can implement a wrapper around the original function pointer to push the default argument to the stack.

## Deliver time and forward compatibility
At model delivery time, two checks can be done:
### Operator check
Two APIs to be provided:
* Runtime: An API to get a runtime’s ops and their schemas (i.e. the # of args). D27920185(WIP)
* Model: An API to get a model’s ops and their schema requirements (i.e. the # of args required).

The APIs can be used to check 
* runtime.ops() is a superset of model.ops()
* for each op in model.ops() validate their schemas are compatible with those in runtime.ops() -- i.e. the # args required in a model op are <= # args in the runtime op.

Note that only root ops in the model needs to be checked here. For transient ops it's not necessary. For example, if a root op, "aten::root" calls "aten::foo", it's "aten::root"'s responsibility to adapt to "aten::foo"'s change, or "aten::root" itself needs to be updated too. 
### Bytecode version backport (PR coming)
When delivering a model with bytecode v6, if the runtime only works with bytecode v5 and lower, backport is needed. 
* The number of arguments is removed from the operator table
* The bytecode version is changed from 6 to 5

Note that this backport is a pure format change, it does not guarantee the backported model always runs in old runtime. The operator check mentioned before should be done first, before it’s back ported to v5.

Differential Revision: [D27986544](https://our.internmc.facebook.com/intern/diff/D27986544)

[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -42,26 +42,43 @@ bool Function::append_operator(
-      stack.push_back(true);
+  // num_specified_args >= 0 indicates number of arguments are available
+  // from model. We can use it to handle backward compatibility.
+  if (num_specified_args >= 0 &&
```
Should this be c10::optional instead? Or does it have some perf cost for model loading?

iseeyuan replied:
Good suggestion! I could use c10::optional to avoid confusion of "-1". I don't this the perf cost for model loading on c10:optional is a big issue.

kimishpatel commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -42,26 +42,43 @@ bool Function::append_operator(
+      auto default_val = args[i].default_value();
+      TORCH_CHECK(
+          default_val.has_value(),
+          "Error happens at preparing for default values for the argument. The ",
```
"Error happened"

kimishpatel commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1165,26 +1165,22 @@ struct MobileCodeImpl : CodeImpl {
+      if (it != op_to_num_specified_args_.end()) {
+        num_include = it->second;
+      }
+      emitLoadInputs(node->inputs(), num_include);
```
WHat is the use of num_include?

iseeyuan replied:
That's the number of specified arguments.

JacobSzwejbka commented on torch/csrc/jit/mobile/function.h:
```diff
@@ -23,7 +23,7 @@ class Function {
       const std::string& name,
       const std::string& overload_name,
-      int64_t model_version);
+      int64_t num_args);
```
nit: should this be num_specified_args

iseeyuan committed:
Update on "[PyTorch Mobile][Forward/backward compatibility] Number of arguments for operators"


Handle forward/backward compatibility caused by added default arguments in mobile. As an example,

In older version, operator aten::foo's schema is
```
foo(Tensor a, Tensor b) -> Tensor
```
In the new version, the schema is updated to
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```

## Model file
Serialize the number of specified arguments to each operator into the bytecode operator table. Before the operator table contains operator name and overload name:
```
('operators', (('aten::foo', ''),))
```
Now the number of specified arguments is added:
```
# bytecode version 6
('operators', (('aten::foo', '', 2),))
```
where "2" means the number of specified arguments. 

Since there's bytecode schema change, the bytecode version number is bumped. This PR is to be landed after #56002 , where the version number is bumped from 4 to 5. This PR bumps the version number from 5 to 6. 

## Runtime and backward compatibility
When the operator is found (either jit or c10), we have the OperatorHandle, where the operator schema can be accessed by 
```
op.value().schema().arguments()
```
Adaptation is implemented to handle backward compatibility. For the example above, the new runtime holds the updated schema: 
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```
Whereas the model file carries
```
(('aten::foo', ''), 2)
```
We can implement a wrapper around the original function pointer to push the default argument to the stack.

## Deliver time and forward compatibility
At model delivery time, two checks can be done:
### Operator check
Two APIs to be provided:
* Runtime: An API to get a runtime’s ops and their schemas (i.e. the # of args). D27920185(WIP)
* Model: An API to get a model’s ops and their schema requirements (i.e. the # of args required).

The APIs can be used to check 
* runtime.ops() is a superset of model.ops()
* for each op in model.ops() validate their schemas are compatible with those in runtime.ops() -- i.e. the # args required in a model op are <= # args in the runtime op.

Note that only root ops in the model needs to be checked here. For transient ops it's not necessary. For example, if a root op, "aten::root" calls "aten::foo", it's "aten::root"'s responsibility to adapt to "aten::foo"'s change, or "aten::root" itself needs to be updated too. 
### Bytecode version backport (PR coming)
When delivering a model with bytecode v6, if the runtime only works with bytecode v5 and lower, backport is needed. 
* The number of arguments is removed from the operator table
* The bytecode version is changed from 6 to 5

Note that this backport is a pure format change, it does not guarantee the backported model always runs in old runtime. The operator check mentioned before should be done first, before it’s back ported to v5.

Differential Revision: [D27986544](https://our.internmc.facebook.com/intern/diff/D27986544)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch Mobile][Forward/backward compatibility] Number of arguments for operators"


Handle forward/backward compatibility caused by added default arguments in mobile. As an example,

In older version, operator aten::foo's schema is
```
foo(Tensor a, Tensor b) -> Tensor
```
In the new version, the schema is updated to
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```

## Model file
Serialize the number of specified arguments to each operator into the bytecode operator table. Before the operator table contains operator name and overload name:
```
('operators', (('aten::foo', ''),))
```
Now the number of specified arguments is added:
```
# bytecode version 6
('operators', (('aten::foo', '', 2),))
```
where "2" means the number of specified arguments. 

Since there's bytecode schema change, the bytecode version number is bumped. This PR is to be landed after #56002 , where the version number is bumped from 4 to 5. This PR bumps the version number from 5 to 6. 

## Runtime and backward compatibility
When the operator is found (either jit or c10), we have the OperatorHandle, where the operator schema can be accessed by 
```
op.value().schema().arguments()
```
Adaptation is implemented to handle backward compatibility. For the example above, the new runtime holds the updated schema: 
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```
Whereas the model file carries
```
(('aten::foo', ''), 2)
```
We can implement a wrapper around the original function pointer to push the default argument to the stack.

## Deliver time and forward compatibility
At model delivery time, two checks can be done:
### Operator check
Two APIs to be provided:
* Runtime: An API to get a runtime’s ops and their schemas (i.e. the # of args). D27920185(WIP)
* Model: An API to get a model’s ops and their schema requirements (i.e. the # of args required).

The APIs can be used to check 
* runtime.ops() is a superset of model.ops()
* for each op in model.ops() validate their schemas are compatible with those in runtime.ops() -- i.e. the # args required in a model op are <= # args in the runtime op.

Note that only root ops in the model needs to be checked here. For transient ops it's not necessary. For example, if a root op, "aten::root" calls "aten::foo", it's "aten::root"'s responsibility to adapt to "aten::foo"'s change, or "aten::root" itself needs to be updated too. 
### Bytecode version backport (PR coming)
When delivering a model with bytecode v6, if the runtime only works with bytecode v5 and lower, backport is needed. 
* The number of arguments is removed from the operator table
* The bytecode version is changed from 6 to 5

Note that this backport is a pure format change, it does not guarantee the backported model always runs in old runtime. The operator check mentioned before should be done first, before it’s back ported to v5.

Differential Revision: [D27986544](https://our.internmc.facebook.com/intern/diff/D27986544)

[ghstack-poisoned]

iseeyuan commented:
@iseeyuan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27986544).

iseeyuan committed:
Update on "[PyTorch Mobile][Forward/backward compatibility] Number of arguments for operators"


Handle forward/backward compatibility caused by added default arguments in mobile. As an example,

In older version, operator aten::foo's schema is
```
foo(Tensor a, Tensor b) -> Tensor
```
In the new version, the schema is updated to
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```

## Model file
Serialize the number of specified arguments to each operator into the bytecode operator table. Before the operator table contains operator name and overload name:
```
('operators', (('aten::foo', ''),))
```
Now the number of specified arguments is added:
```
# bytecode version 6
('operators', (('aten::foo', '', 2),))
```
where "2" means the number of specified arguments. 

Since there's bytecode schema change, the bytecode version number is bumped. This PR is to be landed after #56002 , where the version number is bumped from 4 to 5. This PR bumps the version number from 5 to 6. 

## Runtime and backward compatibility
When the operator is found (either jit or c10), we have the OperatorHandle, where the operator schema can be accessed by 
```
op.value().schema().arguments()
```
Adaptation is implemented to handle backward compatibility. For the example above, the new runtime holds the updated schema: 
```
foo(Tensor a, Tensor b, int groups=1) -> Tensor
```
Whereas the model file carries
```
(('aten::foo', ''), 2)
```
We can implement a wrapper around the original function pointer to push the default argument to the stack.

## Deliver time and forward compatibility
At model delivery time, two checks can be done:
### Operator check
Two APIs to be provided:
* Runtime: An API to get a runtime’s ops and their schemas (i.e. the # of args). D27920185(WIP)
* Model: An API to get a model’s ops and their schema requirements (i.e. the # of args required).

The APIs can be used to check 
* runtime.ops() is a superset of model.ops()
* for each op in model.ops() validate their schemas are compatible with those in runtime.ops() -- i.e. the # args required in a model op are <= # args in the runtime op.

Note that only root ops in the model needs to be checked here. For transient ops it's not necessary. For example, if a root op, "aten::root" calls "aten::foo", it's "aten::root"'s responsibility to adapt to "aten::foo"'s change, or "aten::root" itself needs to be updated too. 
### Bytecode version backport (PR coming)
When delivering a model with bytecode v6, if the runtime only works with bytecode v5 and lower, backport is needed. 
* The number of arguments is removed from the operator table
* The bytecode version is changed from 6 to 5

Note that this backport is a pure format change, it does not guarantee the backported model always runs in old runtime. The operator check mentioned before should be done first, before it’s back ported to v5.

Differential Revision: [D27986544](https://our.internmc.facebook.com/intern/diff/D27986544)

[ghstack-poisoned]

iseeyuan commented:
@iseeyuan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27986544).
